src/retrace/retrace_worker.py

```diff
@@ -641,6 +641,32 @@ def dedup_vmcore(self, md5_task):
 
         return s1.st_size
 
+    def run_crash_cmdline(self, crash_start, crash_cmdline):
+        cmd_output = None
+        returncode = 0
+        try:
+            child = Popen(crash_start, stdin=PIPE, stdout=PIPE, stderr=STDOUT, encoding='utf-8')
```

Comment:
Could't the `encoding` be removed altogether in **this** case? It would get rid of of the decoding error.

_`add_misc()` and `set_atomic()` functions would need to be modified to work with bytes._

Reply:
Personally I don't think that's a good idea since that would make this inconsistent with other calls to Popen in the code.  The way retrace-server is written today, we'll run into this problem on other binaries and commands too if called via Popen with the 'utf-8' encoding and we get some unexpected characters.

Maybe you want to re-think the approach to the original commit that added the 'utf-8' encoding, or add some exception handling?


commit e80d469087ed1118a629bbcaa3d32207b58c8646
Author: Martin Kutlak <mkutlak@redhat.com>
Date:   Fri Jul 27 14:23:47 2018 +0200

    py3-comp: Specify encoding for Popen
    
    If encoding or errors are specified, or text is true, the file objects stdin, stdout and stderr
    are opened in text mode with the specified encoding and errors. The universal_newlines argument
    is equivalent to text and is provided for backwards compatibility. [1]
    
    By default, file objects are opened in binary mode.
    
    encoding argument was added in version 3.6.
    
    [1] https://docs.python.org/3/library/subprocess.html#popen-constructor
    
    Signed-off-by: Martin Kutlak <mkutlak@redhat.com>

Reply:
> Personally I don't think that's a good idea since that would make this inconsistent with other calls to Popen in the code.

I think consistency isn't important with `Popen()` calls.

> Maybe you want to re-think the approach to the original commit that added the 'utf-8' encoding, or add some exception handling?

Yes, it would be good idea to go through the `Popen()` calls and modify cases where the `encoding` isn't needed. 

___
I still think we can remove the `encoding` here. 

As I see it, we run the `crash` command and save its output to a file. There is no string processing involved (no `strip()` or `split()`), so we do not need to decode it to `utf-8`.

Reply:
Ok let me work on that and I'll do some testing and "force push" a new commit.

Reply:
Ok so if you want me to do this then I need to change 'add_misc" as well.  Which means I also need to modify run_gdb() because it too uses 'utf-8' in Popen.  I cannot test this unfortunately can you test it?  If so I can put a more extensive patch series together of multiple patches.

Reply:
>  I cannot test this unfortunately can you test it?

Yes, I can test it.

Reply:
> Ok so if you want me to do this then I need to change 'add_misc" as well.

You also need to change `set_atomic` and `set_backtrace` functions (because of kernellog).

Reply:
Ok I will add a few patches on top of this one because I think that makes the most sense due to all the refactoring.

Reply:
When I remove 'encoding' from Popen I get a TypeError thrown inside 'run_crash_cmdline'
 2019-06-26 13:37:59 crash command: 'log; quit; ' triggered TypeError 
 2019-06-26 13:37:59   memoryview: a bytes-like object is required, not 'str'

This is the code I"m testing:
    def run_crash_cmdline(self, crash_start, crash_cmdline):
        cmd_output = None
        returncode = 0
        try:
            child = Popen(crash_start, stdin=PIPE, stdout=PIPE, stderr=STDOUT)
            cmd_output = child.communicate(crash_cmdline)[0]
        except OSError as err:
            log_warn("crash command: '%s' triggered OSError " %
                     crash_cmdline.replace('\r', '; ').replace('\n', '; '))
            log_warn("  %s" % err)
        except TypeError as err:
            log_warn("crash command: '%s' triggered TypeError " %
                     crash_cmdline.replace('\r', '; ').replace('\n', '; '))
            log_warn("  %s" % err)
        except:
            log_warn("crash command: '%s' triggered Unknown exception %s" %
                     (crash_cmdline.replace('\r', '; ').replace('\n', '; '),
                      sys.exc_info()[0]))

I'm not sure changing to binary opening of stdin/stdout/stderr is the correct approach which is what happens when we remove 'encoding' from Popen.  Normally I think most users would want to know if a crash command output came back with non-text characters (garbage).  I think if we use binary we lose the ability to catch that exception right?

Reply:
Ok I think I figured this out or am getting close.  I might still add an attempt to decode the output and throw an error to the log if we get unusual characters in the crash output but this way we should just dump the output to a file anyway and move on.  Having both the error and the output is best of both worlds.  Let me do some more testing.

Reply:
@mkutlak  is commit 73eaded what you had in mind as far as changes here?
It is not fully done as I've run into some issues with bt_filter.
I've thought for a while these commands would be better in a default, post-retrace hook - the success or fail of a task probably should not depend on them.  However right now may be fine to just refactor and complete this patchset as a first step towards that goal.

Reply:
> @mkutlak is commit [73eaded](https://github.com/abrt/retrace-server/commit/73eadeddc143377e548819d04358dce2be0a6204) what you had in mind as far as changes here?

Yeah, pretty much.

I would just use add a `mode` argument to `add_misc` and `set_backtrace`
```python3
def set_backtrace(self, backtrace, mode="w")

def add_misc(self, name, value, overwrite=False, mode="wb"):
```